### PR TITLE
[Clang] prevent assertion failure by avoiding required literal type checking in C context

### LINKED ISSUE #101304: [Clang][C23] Assertion failure with invalid `constexpr` declaration with incomplete struct
https://godbolt.org/z/q4P5javej

```c++
constexpr struct X x = { 0 };
```

```
test.c:1:20: error: constexpr variable cannot have non-literal type 'const struct X'
    1 | constexpr struct X x = { 0 };
      |                    ^
clang: /llvm/llvm-project/llvm/include/llvm/Support/Casting.h:578: decltype(auto) llvm::cast(From *) [To = clang::CXXRecordDecl, From = clang::RecordDecl]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
0.      Program arguments: /cmake-build-debug/bin/clang -cc1 -std=c23 test.c
1.      test.c:1:22: current parser token '='
 #0 0x00005625b90db92d llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /llvm/llvm-project/llvm/lib/Support/Unix/Signals.inc:723:11
 #1 0x00005625b90dbe1b PrintStackTraceSignalHandler(void*) /llvm/llvm-project/llvm/lib/Support/Unix/Signals.inc:798:1
 #2 0x00005625b90d9e86 llvm::sys::RunSignalHandlers() /llvm/llvm-project/llvm/lib/Support/Signals.cpp:105:5
 #3 0x00005625b90dc5d5 SignalHandler(int) /llvm/llvm-project/llvm/lib/Support/Unix/Signals.inc:413:1
 #4 0x00007f52fc6a1520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #5 0x00007f52fc6f59fc __pthread_kill_implementation ./nptl/./nptl/pthread_kill.c:44:76
 #6 0x00007f52fc6f59fc __pthread_kill_internal ./nptl/./nptl/pthread_kill.c:78:10
 #7 0x00007f52fc6f59fc pthread_kill ./nptl/./nptl/pthread_kill.c:89:10
 #8 0x00007f52fc6a1476 gsignal ./signal/../sysdeps/posix/raise.c:27:6
 #9 0x00007f52fc6877f3 abort ./stdlib/./stdlib/abort.c:81:7
#10 0x00007f52fc68771b _nl_load_domain ./intl/./intl/loadmsgcat.c:1177:9
#11 0x00007f52fc698e96 (/lib/x86_64-linux-gnu/libc.so.6+0x39e96)
#12 0x00005625b94a5078 decltype(auto) llvm::cast<clang::CXXRecordDecl, clang::RecordDecl>(clang::RecordDecl*) /llvm/llvm-project/llvm/include/llvm/Support/Casting.h:579:10
#13 0x00005625bf61a6d7 clang::Sema::RequireLiteralType(clang::SourceLocation, clang::QualType, clang::Sema::TypeDiagnoser&) /llvm/llvm-project/clang/lib/Sema/SemaType.cpp:9270:29
#14 0x00005625bf61b0ac clang::Sema::RequireLiteralType(clang::SourceLocation, clang::QualType, unsigned int) /llvm/llvm-project/clang/lib/Sema/SemaType.cpp:9345:10
#15 0x00005625be1b99a7 clang::Sema::CheckVariableDeclarationType(clang::VarDecl*) /llvm/llvm-project/clang/lib/Sema/SemaDecl.cpp:8759:7
#16 0x00005625be1b6d15 clang::Sema::CheckVariableDeclaration(clang::VarDecl*, clang::LookupResult&) /llvm/llvm-project/clang/lib/Sema/SemaDecl.cpp:8809:7
#17 0x00005625be1b3fc4 clang::Sema::ActOnVariableDeclarator(clang::Scope*, clang::Declarator&, clang::DeclContext*, clang::TypeSourceInfo*, clang::LookupResult&, llvm::MutableArrayRef<clang::TemplateParameterList*>, bool&, llvm::ArrayRef<clang::BindingDecl*>) /llvm/llvm-project/clang/lib/Sema/SemaDecl.cpp:7964:24
#18 0x00005625be1a9a92 clang::Sema::HandleDeclarator(clang::Scope*, clang::Declarator&, llvm::MutableArrayRef<clang::TemplateParameterList*>) /llvm/llvm-project/clang/lib/Sema/SemaDecl.cpp:6405:9
#19 0x00005625be1a8cd0 clang::Sema::ActOnDeclarator(clang::Scope*, clang::Declarator&) /llvm/llvm-project/clang/lib/Sema/SemaDecl.cpp:6046:15
#20 0x00005625bdca55cc clang::Parser::ParseDeclarationAfterDeclaratorAndAttributes(clang::Declarator&, clang::Parser::ParsedTemplateInfo const&, clang::Parser::ForRangeInit*) /llvm/llvm-project/clang/lib/Parse/ParseDecl.cpp:2754:14
#21 0x00005625bdca42fb clang::Parser::ParseDeclGroup(clang::ParsingDeclSpec&, clang::DeclaratorContext, clang::ParsedAttributes&, clang::Parser::ParsedTemplateInfo&, clang::SourceLocation*, clang::Parser::ForRangeInit*) /llvm/llvm-project/clang/lib/Parse/ParseDecl.cpp:2559:9
#22 0x00005625bdbf612c clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec&, clang::AccessSpecifier) /llvm/llvm-project/clang/lib/Parse/Parser.cpp:1249:10
#23 0x00005625bdbf560e clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier) /llvm/llvm-project/clang/lib/Parse/Parser.cpp:1271:12
#24 0x00005625bdbf4e99 clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) /llvm/llvm-project/clang/lib/Parse/Parser.cpp:1074:14
#25 0x00005625bdbf2cbc clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) /llvm/llvm-project/clang/lib/Parse/Parser.cpp:763:12
#26 0x00005625bdbf2360 clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) /llvm/llvm-project/clang/lib/Parse/Parser.cpp:608:8
#27 0x00005625bdbed771 clang::ParseAST(clang::Sema&, bool, bool) /llvm/llvm-project/clang/lib/Parse/ParseAST.cpp:170:15
#28 0x00005625ba6403f1 clang::ASTFrontendAction::ExecuteAction() /llvm/llvm-project/clang/lib/Frontend/FrontendAction.cpp:1194:1
#29 0x00005625ba63fe0c clang::FrontendAction::Execute() /llvm/llvm-project/clang/lib/Frontend/FrontendAction.cpp:1082:7
#30 0x00005625ba55a556 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) /llvm/llvm-project/clang/lib/Frontend/CompilerInstance.cpp:1061:23
#31 0x00005625ba81df3e clang::ExecuteCompilerInvocation(clang::CompilerInstance*) /llvm/llvm-project/clang/lib/FrontendTool/ExecuteCompilerInvocation.cpp:280:8
#32 0x00005625b71dcb29 cc1_main(llvm::ArrayRef<char const*>, char const*, void*) /llvm/llvm-project/clang/tools/driver/cc1_main.cpp:285:13
#33 0x00005625b71cf472 ExecuteCC1Tool(llvm::SmallVectorImpl<char const*>&, llvm::ToolContext const&) /llvm/llvm-project/clang/tools/driver/driver.cpp:215:5
#34 0x00005625b71ce1db clang_main(int, char**, llvm::ToolContext const&) /llvm/llvm-project/clang/tools/driver/driver.cpp:256:5
#35 0x00005625b72038d5 main /cmake-build-debug/tools/clang/tools/driver/clang-driver.cpp:17:3
#36 0x00007f52fc688d90 __libc_start_call_main ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16
#37 0x00007f52fc688e40 call_init ./csu/../csu/libc-start.c:128:20
#38 0x00007f52fc688e40 __libc_start_main ./csu/../csu/libc-start.c:379:5
#39 0x00005625b71cdc65 _start (/cmake-build-debug/clang+0x5bd2c65)
```

It also complains about a "non-literal type", which shouldn't be checked outside of C++

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #101304

### CODE REVIEW

File: clang/lib/Sema/SemaType.cpp [hunk 1]
```diff
@@ -9267,14 +9267,14 @@ bool Sema::RequireLiteralType(SourceLocation Loc, QualType T,
...
 
-  const CXXRecordDecl *RD = cast<CXXRecordDecl>(RT->getDecl());
-
   // A partially-defined class type can't be a literal type, because a literal
   // class type must have a trivial destructor (which can't be checked until
   // the class definition is complete).
   if (RequireCompleteType(Loc, ElemType, diag::note_non_literal_incomplete, T))
     return true;
 
+  const CXXRecordDecl *RD = cast<CXXRecordDecl>(RT->getDecl());
```
Comment (shafik) [inline clang/lib/Sema/SemaType.cpp:9276]: I am surprised we have not run into this before this code was added in 2012.

**Review by shafik — COMMENTED** [review]
This looks good but you need to add a bit more details to your summary, more specifically what case caused the bug to emerge.

I am also curious why this does not show up in C++, we obtain a similar diagnostic but no crash. If we don't have a C++ test similar to this we should add one.

**Comment by MitalAshok** [comment]
I think the issue is here somewhere:

https://github.com/llvm/llvm-project/blob/7088a5ed880f29129ec844c66068e8cb61ca98bf/clang/lib/Sema/SemaDecl.cpp#L8753-L8764

An incomplete struct passes `CheckC23ConstexprVarType` (which it should, incomplete types should fail later) so it goes on to the next check. We should only check `RequireLiteralType` in C++. The cast isn't failing because it's expecting a complete type, it's failing because it's expecting a `CXXRecordDecl` when we only have a C `RecordDecl`.

File: clang/lib/Sema/SemaType.cpp (hunk 1 above)
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaType.cpp:9276]: > I am also curious why this does not show up in C++, we obtain a similar diagnostic but no crash.

Please correct me if I'm wrong, the `CXXRecordDecl` is specific to C++ declarations, so the cast works as expected. In a C env, `RecordDecl` is used instead. We could wrap this in a condition to check specifically for C++, but `*RD` isn't required for `RequireCompleteType`. `RequireCompleteType` handles cases related to incomplete C types, so I'm not sure if we need to perform additional checks here.

```cpp
if (RequireCompleteType(Loc, ElemType, diag::note_non_literal_incomplete, T))
    return true;
```

Comment (a-tarasyuk) [reply clang/lib/Sema/SemaType.cpp:9276]: @MitalAshok Thanks for the feedback. `RequireLiteralType` is called directly in the condition. Should we perform additional checks directly in the condition, or should we include them within `RequireLiteralType` to mark the declaration as invalid?

Comment (MitalAshok) [reply clang/lib/Sema/SemaType.cpp:9276]: You don't require a literal type for C23 `constexpr`, so part of the condition should be `if (getLangOpts().CPlusPlus && NewVD->isConstexpr() && ... && RequireLiteralType`, just like how `CheckC23ConstexprVarType` is gated behind `getLangOpts().C23 &&`. And add an assertion for `getLangOpts().CPlusPlus` in RequireLiteralType for good measure.

The declaration will be marked invalid later by the same code that marks `/*!constexpr*/ struct Incomplete s = {};` invalid.

Comment (a-tarasyuk) [reply clang/lib/Sema/SemaType.cpp:9276]: @MitalAshok Oke, I've added changes. BTW, should I move const `CXXRecordDecl *RD` back to its original position after `RequireCompleteType`? Thanks

Comment (Fznamznon) [reply clang/lib/Sema/SemaType.cpp:9276]: > should I move const CXXRecordDecl *RD back to its original position after RequireCompleteType?

I think yes. It also makes sense to add `assert(LangOpts.CPlusPlus)` in the beginning of `RequireLiteralType` as @MitalAshok already mentioned.

Comment (a-tarasyuk) [reply clang/lib/Sema/SemaType.cpp]: @Fznamznon Thanks. I've reverted. Yep, this isn't necessary to fix the issue., it's more about code cleanup.

**Review by Fznamznon — APPROVED** [review]
LGTM

File: clang/lib/Sema/SemaType.cpp (hunk 1 above)
Comment (a-tarasyuk) [reply clang/lib/Sema/SemaType.cpp]: @shafik @MitalAshok Could you review the latest changes? If there are no further comments, could you please proceed with the merge as I don't have access to do so? Thank you!

Comment (a-tarasyuk) [reply clang/lib/Sema/SemaType.cpp]: @shafik @MitalAshok Could you review the latest changes? If there are no further comments, could you please proceed with the merge as I don't have access to do so? Thank you!

**Review by AaronBallman — APPROVED** [review]
LGTM!

This seems like a pretty reasonable candidate to backport to 19.x as well: https://www.llvm.org/docs/GitHub.html#backporting-fixes-to-the-release-branches

**Pull request merged**